src/authenticator.rs

```diff
@@ -353,8 +349,8 @@ where
             .remove_dir_all(self.options.location, PathBuf::new()))
         .map_err(|_| Status::NotEnoughMemory)?;
 
-        self.state.runtime.reset();
         self._extension_pin_factory_reset()?;
+        self.state.runtime.reset();
```

Comment:
The approach I took in opcard was to call `delete_all` and `remove_dir_all` on all 3 possible locations, that way we can be sure that there is no data left, no possible memory/storage leak. It seems a bit dangerous to rely on the order of the deletion to avoid memory leaks.

Reply:
Can you link it here please?
Like in the volatile storage as well? Yes, that makes sense indeed. I'll check.

Reply:
https://github.com/Nitrokey/opcard-rs/blob/main/src/command.rs#L475

```rust
fn factory_reset<const R: usize, T: trussed::Client>(ctx: Context<'_, R, T>) -> Result<(), Status> {
    *ctx.state = Default::default();
    try_syscall!(ctx
        .backend
        .client_mut()
        .remove_dir_all(Location::Internal, PathBuf::new()))
    .map_err(unspecified_delete_error)?;
    try_syscall!(ctx.backend.client_mut().delete_all(Location::Internal))
        .map_err(unspecified_delete_error)?;
    try_syscall!(ctx
        .backend
        .client_mut()
        .remove_dir_all(Location::External, PathBuf::new()))
    .map_err(unspecified_delete_error)?;
    try_syscall!(ctx.backend.client_mut().delete_all(Location::External))
        .map_err(unspecified_delete_error)?;
    try_syscall!(ctx
        .backend
        .client_mut()
        .remove_dir_all(Location::Volatile, PathBuf::new()))
    .map_err(unspecified_delete_error)?;
    try_syscall!(ctx.backend.client_mut().delete_all(Location::Volatile))
        .map_err(unspecified_delete_error)?;
    Ok(())
}
```

Reply:
What happens if storage on the given location is not existing? Does it not exit early?

Reply:
What do you mean if it does not exists? If it's empty? In that case `remove_dir_all` works fine.

Reply:
This may be possible for the external storage in NFC indeed. Otherwise I think trussed handles creating the directory automatically. It didn't cause issues in opcard tests even though we never use the internal storage (in tests all is done on the external storage).

Reply:
Updated as suggested. Marking resolved.